paimon-core/src/main/java/org/apache/paimon/operation/FileStoreExpireImpl.java

```diff
@@ -226,38 +240,69 @@ private void expireUntil(long earliestId, long endExclusiveId) {
         writeEarliestHint(endExclusiveId);
     }
 
-    private void expireMergeTreeFiles(String manifestListName) {
-        expireMergeTreeFiles(getManifestEntriesFromManifestList(manifestListName));
+    // return a map of partition-buckets of which data files have been deleted
+    private Map<BinaryRow, Set<Integer>> expireMergeTreeFiles(String manifestListName) {
+        return expireMergeTreeFiles(getManifestEntriesFromManifestList(manifestListName));
     }
 
     @VisibleForTesting
-    void expireMergeTreeFiles(Iterable<ManifestEntry> dataFileLog) {
+    Map<BinaryRow, Set<Integer>> expireMergeTreeFiles(Iterable<ManifestEntry> dataFileLog) {
         // we cannot delete a data file directly when we meet a DELETE entry, because that
         // file might be upgraded
-        Map<Path, List<Path>> dataFileToDelete = new HashMap<>();
+        // partition -> bucket -> DataFileMeta
+        Map<BinaryRow, Map<Integer, Set<DataFileMeta>>> dataFileToDelete = new HashMap<>();
+
         for (ManifestEntry entry : dataFileLog) {
-            Path bucketPath = pathFactory.bucketPath(entry.partition(), entry.bucket());
-            Path dataFilePath = new Path(bucketPath, entry.file().fileName());
             switch (entry.kind()) {
                 case ADD:
-                    dataFileToDelete.remove(dataFilePath);
+                    dataFileToDelete.computeIfPresent(
+                            entry.partition(),
+                            (partition, bucketMap) -> {
+                                bucketMap.computeIfPresent(
+                                        entry.bucket(),
+                                        (bucket, files) -> {
+                                            files.remove(entry.file());
```

Comment:
There is bug here, we should add case. See comments above:
```
// we cannot delete a data file directly when we meet a DELETE entry, because that
// file might be upgraded
```